hoodie-common/src/main/java/com/uber/hoodie/common/util/TimelineDiffHelper.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (c) 2019 Uber Technologies, Inc. (hoodie-dev-group@uber.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *          http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.uber.hoodie.common.util;
+
+import com.uber.hoodie.common.table.HoodieTimeline;
+import com.uber.hoodie.common.table.timeline.HoodieInstant;
+import com.uber.hoodie.common.table.timeline.HoodieInstant.State;
+import com.uber.hoodie.common.util.collection.Pair;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+
+
+public class TimelineDiffHelper {
+
+  protected static Logger log = LogManager.getLogger(TimelineDiffHelper.class);
+
+  public static TimelineDiffResult getNewInstantsForIncrementalSync(HoodieTimeline oldTimeline,
+      HoodieTimeline newTimeline) {
+
+    HoodieTimeline oldT = oldTimeline.filterCompletedAndCompactionInstants();
+    HoodieTimeline newT = newTimeline.filterCompletedAndCompactionInstants();
+
+    Optional<HoodieInstant> lastSeenInstant = oldT.lastInstant();
+    Optional<HoodieInstant> firstInstantInNewTimeline = newT.firstInstant();
+
+    if (lastSeenInstant.isPresent() && firstInstantInNewTimeline.isPresent()) {
+      if (HoodieTimeline.compareTimestamps(lastSeenInstant.get().getTimestamp(),
+          firstInstantInNewTimeline.get().getTimestamp(), HoodieTimeline.LESSER)) {
+        // The last seen instant is no longer in the timeline. Do not incrementally Sync.
```

Comment:
I understand that you are concerned about missing some instants in between.. but can't this happen if the commit archival was configured such that, it archived the `lastSeenInstant` and there were new instants after that?

Reply:
Yes, this condition will trigger in the case when lastSeenInstant is archived. We will be conservative and do full-sync.